### PR TITLE
[BEAM-5324] Partially port unpackaged modules to Python 3

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This is is part of a series of PRs with goal to make Apache Beam PY3 compatible. The proposal with the outlined approach has been documented [here](https://s.apache.org/beam-python-3).

Fixing the tests for Python 3 requires a lot of cross-package fixes. I'm therefore submitting this PR with the failing tests skipped during Python 3 testing. This will allow us to build on these fixes while porting the other packages. Since the remaining errors might be caused by other packages, porting of these packages might already solve some of these errors.

In a later stage, these errors should be un-skipped and fixed if they are still failing.

2 tests were skipped in this PR:
- test_side_input_tagged (apache_beam.pipeline_test.DoFnTest)
- test_side_input_no_tag (apache_beam.pipeline_test.DoFnTest)

@tvalentyn @Fematich @charlesccychen @aaltay 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/python/apache_beam/runners/runner.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def create_runner(runner_name):
   if '.' in runner_name:
     module, runner = runner_name.rsplit('.', 1)
     try:
-      return getattr(__import__(module, {}, {}, [runner], -1), runner)()
+      return getattr(__import__(module, {}, {}, [runner], 0), runner)()
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/runner.py]: A note for other reviewers from `__import__` documentation:

> level specifies whether to use absolute or relative imports. The default is -1 which indicates both absolute and relative imports will be attempted. 0 means only perform absolute imports.

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 2]
```diff
@@ -496,7 +496,8 @@ def __init__(self, name_context, spec, counter_factory, state_sampler):
     fn, args, kwargs = pickler.loads(self.spec.combine_fn)[:3]
     self.combine_fn = curry_combine_fn(fn, args, kwargs)
     if (getattr(fn.add_input, 'im_func', None)
-        is core.CombineFn.add_input.__func__):
+        is getattr(core.CombineFn.add_input, '__func__',
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/worker/operations.py]: I am not sure I follow the logic here. Could we instead say:
```
if (fn.add_input.__func__ is core.CombineFn.add_input.__func__):
   ...
```
cc: @robertwb .

Comment (RobbeSneyders) [reply sdks/python/apache_beam/runners/worker/operations.py]: On Python 2, getting a method from a class returns an unbound method, with a reference to the wrapped function stored in `__func__`.
On Pyhon 3, getting a method from a class returns the function itself.

On Python 2.6+, `im_func` is also available as `__func__`, so we can change it to
```Python
if (getattr(fn.add_input, '__func__', None)
    is is getattr(core.CombineFn.add_input, '__func__',
                  core.CombineFn.add_input)):
```
if this is more clear.

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/worker/operations.py]: Thanks for the explanation. Upon a closer look at this code, I think the whole `if` section is legacy code and should be removed.

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/worker/operations.py]: In other words, we always follow `else` branch.

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 3]
```diff
@@ -108,7 +106,7 @@ def __init__(self, operation_name, step_name, consumers, counter_factory,
     self.receivers = [
         operations.ConsumerSet(
             self.counter_factory, self.name_context.step_name, 0,
-            next(itervalues(consumers)), self.windowed_coder)]
+            next(iter(consumers.values())), self.windowed_coder)]
```
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply sdks/python/apache_beam/runners/worker/bundle_processor.py]: You're right. Changed it.

**Comment by tvalentyn** [comment]
LGTM
Thanks, @RobbeSneyders

File: sdks/python/apache_beam/pipeline_test.py [hunk 4]
```diff
@@ -390,6 +391,8 @@ def process(self, element):
     assert_that(pcoll, equal_to([11, 12]))
     pipeline.run()
 
+  @unittest.skipIf(sys.version_info[0] == 3, 'This test still needs to be '
```
Comment (aaltay) [inline sdks/python/apache_beam/pipeline_test.py]: Do we have a JIRA issue to clean up things like this? Should we tag code with proper TODOs?

File: sdks/python/apache_beam/pipeline_test.py [hunk 5]
```diff
@@ -520,13 +525,14 @@ def test_dir(self):
     options = Breakfast()
     self.assertEquals(
         set(['from_dictionary', 'get_all_options', 'slices', 'style',
-             'view_as', 'display_data', 'next']),
-        set([attr for attr in dir(options) if not attr.startswith('_')]))
+             'view_as', 'display_data']),
+        set([attr for attr in dir(options) if not attr.startswith('_') and
+             attr != 'next']))
```
Comment (aaltay) [inline sdks/python/apache_beam/pipeline_test.py]: What is special about 'next' ?

File: sdks/python/apache_beam/runners/runner.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/apache_beam/runners/runner.py]: Could you also explain why is this change needed?

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/runner.py]: I think the reason is the same as why we add `from __future__ import absolute_import` elsewhere - to explicitly disable implicit intra-package import behavior in Python 2 for forward-compatibility with Python 3. @RobbeSneyders, were some tests affected by this?

Related:
http://python-future.org/imports.html
https://www.python.org/dev/peps/pep-0302/

Comment (RobbeSneyders) [reply sdks/python/apache_beam/runners/runner.py]: In Python 3.3+, the -1 option is not available anymore.
Looking at the [_RUNNER_MAP](https://github.com/apache/beam/blob/363d3e385bdb9aca9f194a2b772038bc5db3cb5b/sdks/python/apache_beam/runners/runner.py#L40), all defined runner paths are absolute anyway.

File: sdks/python/apache_beam/pipeline_test.py (hunk 4 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/pipeline_test.py]: Yes, I was cucrrently planning to leave the module specific task open until these skips are resolved.

**Comment by lostluck** [comment]
https://builds.apache.org/job/beam_PostCommit_Python_Verify/6037/ This PR apparently broke the PostCommit (and AFAICT the PreCommit), I'm going to start a revert, to verify.

**Comment by tvalentyn** [comment]
@lostluck There is a fix in flight:  https://github.com/apache/beam/pull/6465

**Comment by lostluck** [comment]
@tvalentyn  Ack. Thank you!

**Comment by tvalentyn** [comment]
FWIW precommit test suite passed before this PR was merged, postcommit was not attempted.

**Pull request merged**